### ISSUE TITLE
Installation issue: slepc-3.12.0 fails to configure with gcc@9.2.0

### ISSUE THREAD

**Issue by albertosalvadori** [spack/spack#13048]
### Steps to reproduce the issue

```console
$ spack install slepc
==> Error: Conflicting conditional dependencies for spec

	slepc@3.12.0%gcc@9.2.0+arpack~blopex arch=linux-centos7-ivybridge ^blas ^gettext ^lapack ^petsc@3.12:3.12.99 ^pkgconfig@0.9.0: ^python@2.6:2.8,3.4:

Cannot merge constraint

	arpack-ng+mpi

into

	arpack-ng~mpi
```

### Platform and user environment

```commandline
$ uname -a 
Linux localhost.localdomain 3.10.0-1062.1.1.el7.x86_64 #1 SMP Fri Sep 13 22:55:44 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
$ lsb_release -d
Description:	CentOS Linux release 7.7.1908 (Core)
``` 

- `packages.yaml`
packages:
  all:
    compiler: [gcc]
  perl:
    paths:
      perl@5.16.3%gcc@9.2.0: /usr
  dealii:
    variants: +optflags~python

- `compilers.yaml`

compilers:
- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules: []
    operating_system: centos7
    paths:
      cc: /usr/bin/gcc
      cxx: /usr/bin/g++
      f77: /usr/bin/gfortran
      fc: /usr/bin/gfortran
    spec: gcc@4.8.5
    target: x86_64
- compiler:
    environment: 
      append-path:
        LD_LIBRARY_PATH: /usr/local/lib64
    extra_rpaths: []
    flags: {}
    modules: []
    operating_system: centos7
    paths:
      cc: /usr/local/bin/gcc
      cxx: /usr/local/bin/g++
      f77: /usr/local/bin/gfortran
      fc: /usr/local/bin/gfortran
    spec: gcc@9.2.0
    target: x86_64

### Additional information

```console
$ less  /tmp/deal.ii/spack-stage/slepc-3.12.0-5md6u45rynyaqtcta4e5dmecqhkp2jkr/spack-src/installed-arch-linux2-c-opt/lib/slepc/conf/configure.log 

================================================================================
Starting Configure Run at Thu Oct  3 20:48:03 2019
Configure Options: --prefix=/home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/slepc-3.12.0-5md6u45rynyaqtcta4e5dmecqhkp2jkr --with-arpack-dir=/home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/arpack-ng-3.7.0-i5fx7mowpxx7acbasidsfc4r3owcd2vx/lib --with-arpack-flags=-lparpack,-larpack
Working directory: /tmp/deal.ii/spack-stage/slepc-3.12.0-5md6u45rynyaqtcta4e5dmecqhkp2jkr/spack-src
Python version:
2.7.16 (default, Oct  3 2019, 20:40:41) 
[GCC 9.2.0]
make: /usr/bin/gmake
PETSc source directory: /home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/petsc-3.12.0-7b3mdm63ap32riorneym2mtcmwjlb63s
PETSc install directory: /home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/petsc-3.12.0-7b3mdm63ap32riorneym2mtcmwjlb63s
PETSc version: 3.12.0
SLEPc source directory: /tmp/deal.ii/spack-stage/slepc-3.12.0-5md6u45rynyaqtcta4e5dmecqhkp2jkr/spack-src
SLEPc install directory: /home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/slepc-3.12.0-5md6u45rynyaqtcta4e5dmecqhkp2jkr
SLEPc version: 3.12.0
================================================================================
Checking PETSc installation...
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
Running command:
cd /tmp/slepc-7TxU8j;/usr/bin/gmake checklink TESTFLAGS=""
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
#include "petscsnes.h"
int main() {
Vec v; Mat m; KSP k;
PetscInitializeNoArguments();
VecCreate(PETSC_COMM_WORLD,&v);
MatCreate(PETSC_COMM_WORLD,&m);
KSPCreate(PETSC_COMM_WORLD,&k);
return 0;
}
/home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/openmpi-3.1.4-4lzhe2gtz3nzhffn6efu2fzgochphcix/bin/mpicc -o checklink.o -c -fPIC   -I/home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/petsc-3.12.0-7b3mdm63ap32riorneym2mtcmwjlb63s/include -I/home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/hypre-2.18.0-dbexk2cnwvnsjd5fm6ltw7o7q66ik3hy/include -I/home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/superlu-dist-6.1.1-stsykz4xojdqtlnjavms2opkppopzush/include -I/home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/hdf5-1.10.5-lt5jyi3ix6dbrbblku7ygutgek7wg5w2/include -I/home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/parmetis-4.0.3-p3vaameiqho6enhkpjhcupk5lam6jvc6/include -I/home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/metis-5.1.0-zepovp3vvqzcirbxoqyb33fg5mm26spe/include -I/home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/zlib-1.2.11-fa7l75havytsbgz77sh6yyzvqgmmm5dj/include    `pwd`/checklink.c
/home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/openmpi-3.1.4-4lzhe2gtz3nzhffn6efu2fzgochphcix/bin/mpicc -fPIC  -o checklink checklink.o  -Wl,-rpath,/home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/petsc-3.12.0-7b3mdm63ap32riorneym2mtcmwjlb63s/lib -L/home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/petsc-3.12.0-7b3mdm63ap32riorneym2mtcmwjlb63s/lib -Wl,-rpath,/home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/hypre-2.18.0-dbexk2cnwvnsjd5fm6ltw7o7q66ik3hy/lib -L/home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/hypre-2.18.0-dbexk2cnwvnsjd5fm6ltw7o7q66ik3hy/lib -Wl,-rpath,/home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/superlu-dist-6.1.1-stsykz4xojdqtlnjavms2opkppopzush/lib -L/home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/superlu-dist-6.1.1-stsykz4xojdqtlnjavms2opkppopzush/lib -Wl,-rpath,/home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/openblas-0.3.7-2dvho7dpjd4bndibkwb7hocq2jcjri3a/lib -L/home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/openblas-0.3.7-2dvho7dpjd4bndibkwb7hocq2jcjri3a/lib -Wl,-rpath,/home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/hdf5-1.10.5-lt5jyi3ix6dbrbblku7ygutgek7wg5w2/lib -L/home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/hdf5-1.10.5-lt5jyi3ix6dbrbblku7ygutgek7wg5w2/lib -Wl,-rpath,/home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/parmetis-4.0.3-p3vaameiqho6enhkpjhcupk5lam6jvc6/lib -L/home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/parmetis-4.0.3-p3vaameiqho6enhkpjhcupk5lam6jvc6/lib -Wl,-rpath,/home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/metis-5.1.0-zepovp3vvqzcirbxoqyb33fg5mm26spe/lib -L/home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/metis-5.1.0-zepovp3vvqzcirbxoqyb33fg5mm26spe/lib -Wl,-rpath,/home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/zlib-1.2.11-fa7l75havytsbgz77sh6yyzvqgmmm5dj/lib -L/home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/zlib-1.2.11-fa7l75havytsbgz77sh6yyzvqgmmm5dj/lib -Wl,-rpath,/home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/hwloc-1.11.11-admmx7jd2wn63yvpnbp4secnkvhpmevu/lib -L/home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/hwloc-1.11.11-admmx7jd2wn63yvpnbp4secnkvhpmevu/lib -Wl,-rpath,/home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/openmpi-3.1.4-4lzhe2gtz3nzhffn6efu2fzgochphcix/lib -L/home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/openmpi-3.1.4-4lzhe2gtz3nzhffn6efu2fzgochphcix/lib -Wl,-rpath,/usr/local/lib/gcc/x86_64-pc-linux-gnu/9.2.0 -L/usr/local/lib/gcc/x86_64-pc-linux-gnu/9.2.0 -Wl,-rpath,/usr/local/lib64 -L/usr/local/lib64 -Wl,-rpath,/usr/local/lib -L/usr/local/lib -lpetsc -lHYPRE -lsuperlu_dist -lopenblas -lhdf5hl_fortran -lhdf5_fortran -lhdf5_hl -lhdf5 -lparmetis -lmetis -lm -lz -lstdc++ -ldl -lmpi_usempif08 -lmpi_usempi_ignore_tkr -lmpi_mpifh -lmpi -lgfortran -lm -lgfortran -lm -lgcc_s -lquadmath -lpthread -lquadmath -lstdc++ -ldl
/home/deal.ii/spack/opt/spack/linux-centos7-ivybridge/gcc-9.2.0/petsc-3.12.0-7b3mdm63ap32riorneym2mtcmwjlb63s/lib/libpetsc.so: undefined reference to `hypre_ParCSRMatrixCompleteClone'
collect2: error: ld returned 1 exit status
gmake: *** [checklink] Error 1

ERROR: Unable to link with PETSc
```

**Comment by davydden** [comment]
@balay may have an idea what's going on here with SLEPc.

**Comment by balay** [comment]
> libpetsc.so: undefined reference to `hypre_ParCSRMatrixCompleteClone'

hypre@2.18.0 had an API change - the fix [add back comatibility]  is in hypre@develop

**Comment by balay** [comment]
> Cannot merge constraint
>	arpack-ng+mpi
> into
>	 arpack-ng~mpi

I've seen this issue for a while. Don't know what triggers this [it doesn't come up with xsdk]. For slepc only build - I've used:

`spack install slepc~arpack`

to avoid hypre issue - I would do:

`spack install slepc~arpack ^hypre@2.17.0`
or
`spack install slepc~arpack ^hypre@develop`

**Comment by davydden** [comment]
> hypre@2.18.0 had an API change

It should be marked as conflict, I guess? Or patched from PETSc package.

**Comment by balay** [comment]
perhaps its better to add the patch in instead of adding the conflict.

https://github.com/spack/spack/pull/13116

cc: @ulrikeyang

**Comment by alalazo** [comment]
Duplicate of #1781

**Comment by alalazo** [comment]
This particular issue was fixed by #13116
